lib/hasPermissions.ts

```diff
@@ -97,6 +103,21 @@ export type HasPermissionsDirectiveArgs = {
 const defaultPolicyOutsideClass: ValidateDirectivePolicy =
   ValidateDirectivePolicy.THROW;
 
+export const getDefaultValue = (
+  container: GraphQLArgument | GraphQLInputObjectType | GraphQLObjectType,
+  path?: Array<string>,
+): unknown => {
+  if ('defaultValue' in container) {
+    return container.defaultValue;
+  }
+  if (isInputObjectType(container) && path && path.length > 0) {
+    const fieldName = path[path.length - 1];
+    const field = container.getFields()[fieldName];
+    return field.defaultValue;
+  }
+  return undefined;
```

Comment:
If no default value is not defined on the schema, it should be checked against null instead of undefined.

e.g.:
using the folowing schema
```graphql
input MyType {
  field: String @hasPermissions(permissions: ["x"])
}
query {
  myQuery(input: MyType!): Boolean
}
```
if the following input is used, the request should be accepted

```graphql
query Q1 {
  myQuery(input: {
    field: null
  })
}
```

```suggestion
  let defaultValue: unknown;
  if ('defaultValue' in container) {
    defaultValue = container.defaultValue;
  }
  if (isInputObjectType(container) && path && path.length > 0) {
    const fieldName = path[path.length - 1];
    const field = container.getFields()[fieldName];
    defaultValue = field.defaultValue;
  }
  return defaultValue ?? null;
```

Reply:
I missed some points on the previous comment.
In some cases, `null` may not be accepted even if sent directly (e.g.: inputs used on mutations), whereas inputs used on queries can accept it without problems.
I think the best way to cover both cases would be adding a new argument to `@hasPermissions`
```graphql
input SomeMutationInput {
  fieldThatCannotReceiveNull: String @hasPermissions(permissions: ["x"])
  fieldThatCanReceiveNull: String @hasPermissions(permissions: ["x"], noDefaultEqualsNull: true)
}
input InputWhoseFieldsCannotReceiveNullWithoutPermission @hasPermissions(permissions: ["x"]){
  inputField: String
}
input SomeQueryInput @hasPermissions(permissions: ["y"], noDefaultEqualsNull: true) {
  fieldThatCanReceiveNull: String
}
mutation {
  updateSomething(input: SomeMutationInput!, otherInput: InputWhoseFieldsCannotReceiveNullWithoutPermission): String
}
query {
  someQuery(input: SomeQueryInput): Boolean
}
```
This way, one can choose which inputs allows explicitly sending null when defining the schema.
Using the example schema above and an user that doesn't have the necessary permissions to send the fields, the following should work:
```graphql
query myQuery {
  someQuery(input: {
    fieldThatCanReceiveNull: null
  })
}

mutation myMutation {
  updateSomething(input: {
    fieldThatCanReceiveNull: null
  })
}
```
And the following shouldn't
```graphql
mutation myMutation {
  updateSomething(input: {
    fieldThatCannotReceiveNull: null
  })

  updateSomething(otherInput: {
    inputField: null
  })
}
```

Note that for fields that have objects, checking this is more complex, but using `lodash.isEqualWith` may suffice.
e.g.:
```typescript
const compareWithDefaultAsNull = (value: unknown, default: unknown): boolean | undefined => {
  if(default === undefined){
    return value === null;
  }
  // delegate to underlying isEqual
  return undefined
}
const compareWithDefaultAsIs = (_value: unknown, _default: unknown): undefined => undefined

//...
let compareFn = compareWithDefaultAsIs;
if(noDefaultEqualsNull){
  compareFn = compareWithDefaultAsNull 
}
if(isEqualWith(value,  defaultValue, compareFn)){
  return value
}
// checkPermissions
```

Reply:
@DDDKnightmare arguments shouldn't have any difference if they are used on mutations or queries, I'd avoid messing with that as you propose, instead just check for undefined and that's it. No extra arguments or the likes. The schema may declare `= null` as a default value if that's desired.

We need to reduce the variants and make it more consistent, your proposed change will make it less consistent and less predictable.